File: members/U02BWDJEB2Q/programmers/괄호_변환.java
package programmers;

import java.util.Stack;

public class 괄호_변환 {

    public static void main(String[] args) {
        System.out.println(solution("(()())()"));
        System.out.println(solution(")("));
        System.out.println(solution("()))((()"));
    }

    public static String solution(String p) {
        if (p.length() == 0) {
            return "";
        }

        int countOfLeft = 0;
        int countOfRight = 0;

        int i;

        for (i = 0; i < p.length(); i++) {
            if (p.charAt(i) == '(') {
                countOfLeft += 1;
            } else {
                countOfRight += 1;
            }

            if (countOfLeft == countOfRight) {
                break;
            }
        }

        String u = p.substring(0, i + 1);
        String v = i + 1 < p.length() ? p.substring(i + 1) : "";

        if (isCorrect(u)) {
            u += solution(v);
        }

        if (!isCorrect(u)) {
            StringBuilder result = new StringBuilder("(");

            result.append(solution(v));

            result.append(")");

            u = u.substring(1, u.length() - 1);

            for (i = 0; i < u.length(); i++) {
                result.append(u.charAt(i) == '(' ? ')' : '(');
            }

            return result.toString();
        }

        return u;
    }

    private static boolean isCorrect(String u) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < u.length(); i++) {
            char c = u.charAt(i);

            if (c == '(') {
                stack.push(c);
            } else {
                if (!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                } else {
                    stack.push(c);
                }
            }
        }

        return stack.isEmpty();
    }
}

File: members/U02C3AE43C1/programmers/괄호 변환.js
function trans(str) {
  if (str === '') return '';

  let u = '',
    v = '';

  for (let i = 0, lCnt = 0, rCnt = 0; i < str.length; i++) {
    if (str[i] === '(') {
      lCnt += 1;
    } else {
      rCnt += 1;
    }

    if (lCnt === rCnt) {
      u = str.slice(0, i + 1);
      v = str.slice(i + 1);
      break;
    }
  }

  if (u[0] === '(') {
    return u + trans(v);
  }

  u = [...u.slice(1, u.length - 1)].map((char) => (char === '(' ? ')' : '(')).join('');

  return '(' + trans(v) + ')' + u;
}

function solution(p) {
  return trans(p);
}

// function trans(str) {
//     if(str === '') return '';

//     let u = '', v = '';

//     for(let i = 0, lCnt = 0, rCnt = 0;i < str.length;i++) {
//         if(str[i] === '(') {
//             lCnt += 1;
//         } else {
//             rCnt += 1;
//         }

//         if(lCnt === rCnt) {
//             u = str.slice(0, i + 1);
//             v = str.slice(i + 1);
//             break;
//         }
//     }

//     if(isCorrect(u)) return u + trans(v);

//     u = [...u.slice(1, u.length - 1)]
//         .map(char => char === '(' ? ')' : '(')
//         .join('');

//     return '(' + trans(v) + ')' + u;
// }

// function isCorrect(str) {
//     const stack = [];

//     [...str].forEach(char => {
//         if(stack.length === 0) {
//             stack.push(char);
//         } else if(char === '(') {
//             stack.push(char);
//         } else {
//             if(stack[stack.length - 1] === '(') {
//                 stack.pop();
//             } else {
//                 stack.push(char);
//             }
//         }
//     });

//     return stack.length === 0;
// }

// function solution(p) {
//     return trans(p);
// }
